src/BitStringAddresses/fockaddress.jl

```diff
@@ -388,11 +402,250 @@ This function is used to implement `Base.show` for [`AbstractFockAddress`](@ref)
 print_address
 
 function Base.show(io::IO, add::AbstractFockAddress)
-    if get(io, :compact, false)
+    if get(io, :typeinfo, nothing) == typeof(add) || get(io, :compact, false)
         print(io, "fs\"")
-        print_address(io, add)
+        print_address(io, add; compact=true)
         print(io, "\"")
     else
-        print_address(io, add)
+        print_address(io, add; compact=false)
+    end
+end
+
+function onr_sparse_string(o)
+    ps = map(p -> p[1] => p[2], Iterators.filter(p -> !iszero(p[2]), enumerate(o)))
+    return join(ps, ", ")
+end
+
+###
+### Boson stuff
+###
+"""
+    BoseFSIndex
+
+Struct used for indexing and performing [`excitation`](@ref)s on a [`BoseFS`](@ref).
+
+## Fields:
+
+* `occnum`: the occupation number.
+* `mode`: the index of the mode.
+* `offset`: the position of the mode in the address. This is the bit offset of the mode when
+ the address is represented by a bitstring, and the position in the list when it is
+ represented by `SortedParticleList`.
+
+"""
+struct BoseFSIndex<:FieldVector{3,Int}
+    occnum::Int
+    mode::Int
+    offset::Int
+end
+
+function Base.show(io::IO, i::BoseFSIndex)
+    @unpack occnum, mode, offset = i
+    print(io, "BoseFSIndex(occnum=$occnum, mode=$mode, offset=$offset)")
+end
+Base.show(io::IO, ::MIME"text/plain", i::BoseFSIndex) = show(io, i)
+
+"""
+    BoseOccupiedModes{C,S<:BoseFS}
+
+Iterator for occupied modes in [`BoseFS`](@ref). The definition of `iterate` is dispatched
+on the storage type.
+
+See [`occupied_modes`](@ref).
+
+Defining `Base.length` and `Base.iterate` for this struct is a part of the interface for an
+underlying storage format used by [`BoseFS`](@ref).
+"""
+struct BoseOccupiedModes{N,M,S}<:OccupiedModeIterator
+    storage::S
+end
+Base.eltype(::BoseOccupiedModes) = BoseFSIndex
+
+# Apply destruction operator to BoseFSIndex.
+@inline _destroy(d, index::BoseFSIndex) = @set index.occnum -= (d.mode == index.mode)
+@inline _destroy(d) = Base.Fix1(_destroy, d)
+# Apply creation operator to BoseFSIndex.
+@inline _create(c, index::BoseFSIndex) = @set index.occnum += (c.mode == index.mode)
+@inline _create(c) = Base.Fix1(_create, c)
+
+"""
+    bose_excitation_value(
+        creations::NTuple{_,BoseFSIndex}, destructions::NTuple{_,::BoseFSIndex}
+    ) -> Float64
```

Comment:
```suggestion
    bose_excitation_value(
        creations::NTuple{_,BoseFSIndex}, destructions::NTuple{_,::BoseFSIndex}
    ) -> Int64
```
Is this correct? Just from reading the code it looks like this function is returning an integer, which is actually the square of the excitation amplitude. If this is true, then maybe it could be clarified in the docstring.